File: backend/src/main/java/com/back/back9/global/problem/ProblemDetails.java
package com.back.back9.global.problem;

import com.back.back9.global.error.ErrorCode;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;

import java.net.URI;
import java.time.OffsetDateTime;
import java.util.Map;

public final class ProblemDetails {

    private ProblemDetails() {}

    public static ProblemDetail of(ErrorCode code, Object... args) {
        String detail = args == null || args.length == 0
                ? code.getDefaultDetail()
                : String.format(code.getDefaultDetail(), args);


        ProblemDetail pd = ProblemDetail.forStatus(code.getStatus());
        pd.setTitle(code.name());
        pd.setType(URI.create("about:blank"));
        pd.setProperty("code", code.getCode());
        pd.setProperty("timestamp", OffsetDateTime.now());
        return pd;

    }

    public static ProblemDetail of(HttpStatus status, String title,
                                   String detail,
                                   String code,
                                   Map<String, Object> properties) {
        ProblemDetail pd = ProblemDetail.forStatusAndDetail(status, detail);
        pd.setTitle(title);

        if(code != null) pd.setProperty("code", code);
        if(properties != null) {
            properties.forEach(pd::setProperty);
        }
        pd.setProperty("timestamp", OffsetDateTime.now());

        return pd;
}
}

File: backend/src/main/kotlin/com/back/back9/global/problem/ProblemDetails.kt
package com.back.back9.global.problem

import com.back.back9.global.error.ErrorCode
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import java.net.URI
import java.time.OffsetDateTime

object ProblemDetails {

    fun of(code: ErrorCode, vararg args: Any?): ProblemDetail {
        val detail = if (args.isEmpty()) code.defaultDetail
        else String.format(code.defaultDetail, *args)

        val pd = ProblemDetail.forStatus(code.status)
        pd.title = code.name
        pd.type = URI.create("about:blank")
        pd.setProperty("code", code.code)
        pd.setProperty("timestamp", OffsetDateTime.now())
        // Java 원본과 동일하게 detail은 설정하지 않음(로직 보존)
        return pd
    }

    fun of(
        status: HttpStatus,
        title: String,
        detail: String,
        code: String?,
        properties: Map<String, Any>?
    ): ProblemDetail {
        val pd = ProblemDetail.forStatusAndDetail(status, detail)
        pd.title = title
        if (code != null) pd.setProperty("code", code)
        properties?.forEach { (k, v) -> pd.setProperty(k, v) }
        pd.setProperty("timestamp", OffsetDateTime.now())
        return pd
    }
}
